types/src/share.rs

```diff
@@ -0,0 +1,172 @@
+use celestia_proto::share::p2p::shrex::nd::{Proof as RawProof, Row as RawRow};
+use nmt_rs::simple_merkle::proof::Proof as NmtProof;
+use serde::{Deserialize, Serialize};
+use tendermint_proto::Protobuf;
+
+use crate::nmt::{Namespace, NamespaceProof, NamespacedHash, NS_SIZE};
+use crate::{Error, Result};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct NamespacedShares {
+    pub rows: Vec<NamespacedRow>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(try_from = "RawRow", into = "RawRow")]
+pub struct NamespacedRow {
+    pub shares: Vec<Share>,
+    pub proof: NamespaceProof,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Share {
+    pub namespace: Namespace,
+    pub data: Vec<u8>,
+}
+
+impl Share {
+    fn new(bytes: Vec<u8>) -> Result<Self> {
+        if bytes.len() < NS_SIZE {
+            return Err(Error::InvalidShareSize(bytes.len()));
+        }
+
+        let (ns, data) = bytes.split_at(NS_SIZE);
+
+        Ok(Share {
+            namespace: Namespace::from_raw(ns)?,
+            data: data.to_vec(),
+        })
+    }
+
+    fn _to_vec(&self) -> Vec<u8> {
+        let mut bytes = self.namespace.as_bytes().to_vec();
+        bytes.extend_from_slice(&self.data);
+        bytes
+    }
```

Comment:
shouldn't this (and `new`) be public instead?

Reply:
*instead of `_` prefix

Reply:
The main reason that `new` is not public is because the user will never construct it manually from raw bytes. Construction will be done only from `RawRow`. I added `_` prefix on `to_vec` intetionally since it was dead code and it is something that we will use in `impl From<NamespacedRow> for RawRow`. Another option was to add `#[allow(dead_code)]`. I am ok with both of them.